// ataraxia-transport/dist/cjs/auth/AnonymousAuth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnonymousAuth = void 0;
const AuthServerFlow_1 = require("./AuthServerFlow");
/**
 * Anonymous authentication. Allows anyone to connect to the network without
 * any form of authentication.
 */
class AnonymousAuth {
    constructor() {
        this.id = 'anonymous';
    }
    createClientFlow() {
        return {
            initialMessage() {
                return Promise.resolve(new ArrayBuffer(0));
            },
            receiveData() {
                return Promise.reject();
            },
            destroy() {
                return Promise.resolve();
            }
        };
    }
    createServerFlow() {
        return {
            receiveInitial(data) {
                if (data.byteLength !== 0) {
                    return Promise.resolve({
                        type: AuthServerFlow_1.AuthServerReplyType.Reject
                    });
                }
                return Promise.resolve({
                    type: AuthServerFlow_1.AuthServerReplyType.Ok
                });
            },
            receiveData() {
                return Promise.reject();
            },
            destroy() {
                return Promise.resolve();
            }
        };
    }
}
exports.AnonymousAuth = AnonymousAuth;
AnonymousAuth.INSTANCE = new AnonymousAuth();
