// ataraxia-transport/dist/esm/DisconnectReason.js
/**
 * Reason for a disconnect. Used to flag why a peer was disconnected so a
 * transport can take different actions such as reconnecting after a delay or
 * directly.
 */
export var DisconnectReason;
(function (DisconnectReason) {
    /**
     * Manual disconnect has happened.
     */
    DisconnectReason[DisconnectReason["Manual"] = 0] = "Manual";
    /**
     * A generic error occurred while talking to the peer.
     */
    DisconnectReason[DisconnectReason["Error"] = 1] = "Error";
    /**
     * The peer didn't respond to a ping in time.
     */
    DisconnectReason[DisconnectReason["PingTimeout"] = 2] = "PingTimeout";
    /**
     * Negotiation failed. Could not negotiate with the peer about how to
     * communicate. This is either due to the peer not responding in time, or
     * it responding in an unexpected way.
     */
    DisconnectReason[DisconnectReason["NegotiationFailed"] = 3] = "NegotiationFailed";
    /**
     * Authentication was rejected. Either the peer rejected our authentication
     * or we rejected the peers authentication.
     */
    DisconnectReason[DisconnectReason["AuthReject"] = 4] = "AuthReject";
})(DisconnectReason || (DisconnectReason = {}));
