// ataraxia-transport/dist/esm/auth/AnonymousAuth.js
import { AuthServerReplyType } from './AuthServerFlow';
/**
 * Anonymous authentication. Allows anyone to connect to the network without
 * any form of authentication.
 */
export class AnonymousAuth {
    constructor() {
        this.id = 'anonymous';
    }
    createClientFlow() {
        return {
            initialMessage() {
                return Promise.resolve(new ArrayBuffer(0));
            },
            receiveData() {
                return Promise.reject();
            },
            destroy() {
                return Promise.resolve();
            }
        };
    }
    createServerFlow() {
        return {
            receiveInitial(data) {
                if (data.byteLength !== 0) {
                    return Promise.resolve({
                        type: AuthServerReplyType.Reject
                    });
                }
                return Promise.resolve({
                    type: AuthServerReplyType.Ok
                });
            },
            receiveData() {
                return Promise.reject();
            },
            destroy() {
                return Promise.resolve();
            }
        };
    }
}
AnonymousAuth.INSTANCE = new AnonymousAuth();
